### PR TITLE
Make btl/usnic thread safe.

### PR DESCRIPTION
**Pull request by thananon**
MPI_THREAD_MULTIPLE support for btl/usnic.

### CODE REVIEW

File: opal/mca/btl/usnic/btl_usnic_module.h [hunk 1]
```diff
@@ -54,6 +54,12 @@
 BEGIN_C_DECLS
 
 /*
+ * MPI_THREAD_MULTIPLE support
+ */
+opal_recursive_mutex_t btl_usnic_send_lock;
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.h]: Need an extern here.

File: opal/mca/btl/usnic/btl_usnic_module.h [hunk 2]
```diff
@@ -251,7 +257,7 @@ opal_btl_usnic_get_next_endpoint_needing_ack(
 static inline void
 opal_btl_usnic_remove_from_endpoints_needing_ack(
     opal_btl_usnic_endpoint_t *endpoint)
-{
+{
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.h]: Extra whitespace!

File: opal/mca/btl/usnic/btl_usnic_recv.h [hunk 3]
```diff
@@ -157,8 +156,10 @@ opal_btl_usnic_check_rx_seq(
 #if MSGDEBUG1
         opal_output(0, "Handle piggy-packed ACK seq %"UDSEQ"\n", seg->rs_base.us_btl_header->ack_seq);
 #endif
+    OPAL_THREAD_LOCK(&btl_usnic_send_lock);
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_recv.h]: Wrong indenting.

File: opal/mca/btl/usnic/btl_usnic_send.h [hunk 4]
```diff
@@ -264,7 +265,7 @@ opal_btl_usnic_endpoint_enqueue_frag(
     opal_btl_usnic_endpoint_t *endpoint,
     opal_btl_usnic_send_frag_t *frag)
 {
-#if MSGDEBUG1
+#if MSGDEBUG1
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_send.h]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_send.h [hunk 5]
```diff
@@ -253,6 +253,7 @@ opal_btl_usnic_endpoint_send_segment(
        endpoint hotel */
     opal_hotel_checkin_with_res(&endpoint->endpoint_hotel, sseg,
             &sseg->ss_hotel_room);
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_send.h]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_module.c [hunk 6]
```diff
@@ -1754,7 +1760,7 @@ static int init_one_channel(opal_btl_usnic_module_t *module,
     if (OPAL_SUCCESS != rc) {
         goto error;
     }
-
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.c]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_module.c [hunk 7]
```diff
@@ -1122,6 +1125,7 @@ opal_btl_usnic_module_progress_sends(
             opal_list_remove_item(&endpoint->endpoint_frag_send_queue,
                     &frag->sf_base.uf_base.super.super);
 
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.c]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_module.c [hunk 8]
```diff
@@ -1104,6 +1106,7 @@ opal_btl_usnic_module_progress_sends(
          * the endpoint's send window is open, and the endpoint has send
          * credits.
          */
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.c]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_module.c [hunk 9]
```diff
@@ -967,6 +967,7 @@ usnic_do_resends(
             abort();    /* should not be possible */
         }
     }
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.c]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_module.c [hunk 10]
```diff
@@ -23,6 +23,7 @@
  * $HEADER$
  */
 
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.c]: Extra whitespace

File: opal/mca/btl/usnic/btl_usnic_component.c [hunk 11]
```diff
@@ -618,10 +618,11 @@ static mca_btl_base_module_t** usnic_component_init(int* num_btl_modules,
     /* Currently refuse to run if MPI_THREAD_MULTIPLE is enabled */
     if (want_mpi_threads && !mca_btl_base_thread_multiple_override) {
         opal_output_verbose(5, USNIC_OUT,
-                            "btl:usnic: MPI_THREAD_MULTIPLE not supported; skipping this component");
-        return NULL;
+                            "btl:usnic: MPI_THREAD_MULTIPLE support is in testing phase.");
     }
 
+    OBJ_CONSTRUCT(&btl_usnic_send_lock, opal_recursive_mutex_t);
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_component.c]: Is there a corresponding destruct?

File: opal/mca/btl/usnic/btl_usnic_ack.c [hunk 12]
```diff
@@ -90,6 +90,7 @@ opal_btl_usnic_handle_ack(
 
     /* Does this ACK have a new sequence number that we haven't
        seen before? */
+    sseg = NULL;
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_ack.c]: Make this a separate commit, please.

File: opal/mca/btl/usnic/README.txt [hunk 13]
```diff
@@ -335,3 +335,40 @@ libfabric abstractions:
...
+
+With this concept, many functions in btl/usnic that make calls to the
+listed functions are protected by OPAL_THREAD_LOCK macro which will only 
+be active if the user specify MPI_Init_thread() with MPI_THREAD_MULTIPLE 
+support.
+
+
+
+       
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/README.txt]: No blank lines at end of file, please.  😄

File: opal/mca/btl/usnic/btl_usnic_recv.c [hunk 14]
```diff
@@ -340,8 +340,9 @@ void opal_btl_usnic_recv_call(opal_btl_usnic_module_t *module,
         opal_output(0, "    Received ACK for sequence number %" UDSEQ " from %s to %s\n",
                     bseg->us_btl_header->ack_seq, remote_ip, local_ip);
 #endif
+        OPAL_THREAD_LOCK(&btl_usnic_send_lock);
         opal_btl_usnic_handle_ack(endpoint, ack_seq);
-
+        OPAL_THREAD_UNLOCK(&btl_usnic_send_lock);
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_recv.c]: It's a bit odd that the lock is named "send_lock" when you also use it in the receive.  I'm guessing this is left over from when you had multiple locks.  Can you rename the lock to be something more generic -- perhaps even `btl_usnic_lock`?

File: opal/mca/btl/usnic/README.txt [hunk 15]
```diff
@@ -335,3 +335,35 @@ libfabric abstractions:
...
+- opal_btl_usnic_check_rts()  
+- opal_btl_usnic_handle_ack()  
+
+We also have to protect the call to libfabric routines
+
+- opal_btl_usnic_endpoint_send_segment()        (fi_send)
+- opal_btl_usnic_recv_call()			(fi_recvmsg)
+
+have to be protected as well.
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/README.txt:365]: Don't forget to mention the cclient usage has to be protected (but not the cagent).

File: opal/mca/btl/usnic/btl_usnic_module.c [hunk 16]
```diff
@@ -1166,6 +1167,7 @@ opal_btl_usnic_module_progress_sends(
         /* If no more sends or endpoint send window is closed,
          * or no more send credits, remove from send list
          */
+
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_module.c]: Extra whitespace.

File: opal/mca/btl/usnic/btl_usnic_component.c [hunk 17]
```diff
@@ -618,10 +623,11 @@ static mca_btl_base_module_t** usnic_component_init(int* num_btl_modules,
     /* Currently refuse to run if MPI_THREAD_MULTIPLE is enabled */
     if (want_mpi_threads && !mca_btl_base_thread_multiple_override) {
         opal_output_verbose(5, USNIC_OUT,
-                            "btl:usnic: MPI_THREAD_MULTIPLE not supported; skipping this component");
-        return NULL;
+                            "btl:usnic: MPI_THREAD_MULTIPLE support is in testing phase.");
```
Comment (jsquyres) [inline opal/mca/btl/usnic/btl_usnic_component.c]: We did not talk about this part, but the whole reason the `_compat` code exists in the usNIC BTL is so that we can have the same code base in both the v1.10 and v2.x series.

Since we really only care about THREAD_MULTIPLE support in 2.x, you should probably put a conditional here:
- For versions < 2.0.0, do the old behavior (return NULL and the original verbose message)
- For versions >= 2.0.0, do the new behavior

**Comment by jsquyres** [comment]
Once you fix up these minor things, I think you're good to go to:
- Squash into a single commit (with a nice descriptive commit message)
- Make a PR against ompi master

File: opal/mca/btl/usnic/btl_usnic_recv.c (hunk 14 above)
Comment (thananon) [inline opal/mca/btl/usnic/btl_usnic_recv.c]: Yes, I changed it locally for a while already. Will push with the rest.
